File: tests/bindings/cpp/memory_backend_fixture.cpp
#include "common/backend.hpp"
#include "common/memory/backend.hpp"
#include "common/runtime/feature_flag_store.hpp"
#include "common/runtime/lease_store.hpp"
#include "common/runtime/log_sink.hpp"
#include "common/runtime/metric_sink.hpp"
#include "common/runtime/queue_store.hpp"
#include "common/runtime/workflow_store.hpp"

#include <cassert>
#include <chrono>

int main()
{
    using namespace statespec::backend;
    using namespace statespec::backend::memory;
    using namespace statespec::backend::runtime;

    InMemoryBackend backend;
    RuntimeFeatureFlagStore flags;
    RuntimeQueueStore queues;
    RuntimeLeaseStore leases;
    RuntimeWorkflowStore workflows;
    RuntimeLogSink logs;
    RuntimeMetricSink metrics;

    backend.ensure_collection(CollectionDescriptor{.name = "orders", .key_fields = {"order_id"}});

    auto tx = backend.begin();
    flags.register_definitionTx(
        *tx, FeatureFlagDefinition{
                 .name = "new_scheduler",
                 .type = FeatureFlagType::Bool,
                 .default_value = FeatureFlagValue::bool_value(true),
                 .scope = FeatureFlagScopeKind::System,
             }
    );
    queues.register_definitionTx(
        *tx, RegisterQueueDefinitionRequest{
                 .definition = QueueDefinition{
                     .queue = "workflow",
                     .channel = "launch",
                     .visibility_timeout = std::chrono::minutes{1},
                     .max_attempts = 3,
                     .metadata = Json::object({}),
                 },
             }
    );

    LeaseDefinitionId lease_id{.name = "workflow_step", .version = 1};
    leases.register_definitionTx(
        *tx, LeaseDefinition{
                 .id = lease_id,
                 .resource_pattern = "workflow/*",
                 .ttl = std::chrono::minutes{1},
                 .renew_every = std::chrono::seconds{30},
                 .fencing_token = true,
             }
    );
    workflows.register_definitionTx(
        *tx, RegisterWorkflowDefinitionRequest{
                 .definition = WorkflowDefinition{
                     .workflow_name = "ProvisionService",
                     .workflow_version = 1,
                     .start_step = "validate_request",
                     .expected_execution_time = std::chrono::minutes{1},
                     .steps =
                         {
                             WorkflowStepDefinition{
                                 .name = "validate_request",
                                 .expected_execution_time = std::chrono::seconds{1},
                                 .max_retries = 3,
                             },
                         },
                     .metadata = Json::object({}),
                 },
             }
    );
    logs.register_definitionTx(
        *tx, LogDefinition{
                 .name = "launch_decision",
                 .level = LogLevel::Info,
                 .event_name = "workflow.launch.decision",
             }
    );
    metrics.register_definitionTx(
        *tx, MetricDefinition{
                 .name = "launch_attempts",
                 .kind = MetricKind::Counter,
                 .backend_name = "workflow_launch_attempts_total",
                 .unit = "1",
             }
    );
    backend.commit(*tx);

    assert(
        flags.evaluate(backend, FeatureFlagEvaluationRequest{.name = "new_scheduler"}).as_bool() ==
        true
    );

    const auto now = std::chrono::system_clock::from_time_t(100);
    auto message = queues.enqueue(
        backend, EnqueueMessageRequest{
                     .message_id = "msg-1",
                     .queue = "workflow",
                     .channel = "launch",
                     .payload = Json::object({}),
                 }
    );
    assert(message.status == "Pending");

    auto claimed = queues.claim(
        backend, ClaimMessageRequest{
                     .queue = "workflow",
                     .channel = "launch",
                     .claimant = "worker-1",
                     .now = now,
                     .visibility_timeout = std::chrono::minutes{1},
                     .max_messages = 1,
                 }
    );
    assert(claimed.size() == 1);

    auto lease = leases.acquire(
        backend, LeaseAcquireRequest{
                     .definition_id = lease_id,
                     .resource = "workflow/msg-1",
                     .holder = "worker-1",
                     .now = now,
                 }
    );
    assert(lease.acquired);
    assert(lease.lease.has_value());

    auto started = workflows.start(
        backend, StartWorkflowRequest{
                     .workflow_execution_id = "wf-1",
                     .workflow_name = "ProvisionService",
                     .workflow_version = 1,
                     .start_step = "validate_request",
                     .state = Json::object({}),
                 }
    );
    assert(started.status == "Running");

    auto steps = workflows.claim_steps(
        backend, ClaimWorkflowStepRequest{
                     .workflow_execution_id = "wf-1",
                     .workflow_name = "ProvisionService",
                     .workflow_version = 1,
                     .worker = "worker-1",
                     .now = now,
                     .lease_duration = std::chrono::minutes{1},
                     .max_steps = 1,
                 }
    );
    assert(steps.size() == 1);

    workflows.keep_alive_step(
        backend, KeepAliveWorkflowStepRequest{
                     .workflow_execution_id = "wf-1",
                     .worker = "worker-1",
                     .current_step = "validate_request",
                     .now = now,
                     .lease_duration = std::chrono::minutes{1},
                 }
    );

    logs.emit_log(
        backend, LogEvent{
                     .name = "launch_decision",
                     .level = LogLevel::Info,
                     .event_name = "workflow.launch.decision",
                 }
    );
    metrics.record_metric(
        backend, MetricSample{
                     .name = "launch_attempts",
                     .kind = MetricKind::Counter,
                     .backend_name = "workflow_launch_attempts_total",
                     .value = 1.0,
                     .unit = "1",
                 }
    );
    assert(logs.inspect_events(backend).size() == 1);
    assert(metrics.inspect_samples(backend).size() == 1);

    auto write_tx = backend.begin();
    backend.put(*write_tx, "orders", "order-1", Json::object({{"status", "new"}}));
    backend.commit(*write_tx);

    auto read_tx = backend.begin();
    assert(backend.query(*read_tx, "orders", Query::key_prefix_query("order-")).size() == 1);
    backend.commit(*read_tx);

    return 0;
}

File: tests/bindings/rust/memory_backend_fixture.rs
use std::collections::BTreeMap;
use std::time::{Duration, SystemTime};

use statespec_generated::backend::{Backend, Query};
use statespec_generated::feature_flag::{
    FeatureFlagDefinition, FeatureFlagEvaluationContext, FeatureFlagEvaluationRequest,
    FeatureFlagScopeKind, FeatureFlagStore, FeatureFlagType, FeatureFlagValue,
};
use statespec_generated::json::Json;
use statespec_generated::lease::{
    LeaseAcquireRequest, LeaseDefinition, LeaseDefinitionId, LeaseStore,
};
use statespec_generated::log::{LogDefinition, LogEvent, LogLevel, LogSink};
use statespec_generated::memory_backend::InMemoryBackend;
use statespec_generated::metric::{MetricDefinition, MetricKind, MetricSample, MetricSink};
use statespec_generated::queue::{
    ClaimMessageRequest, EnqueueMessageRequest, QueueDefinition, QueueStore,
    RegisterQueueDefinitionRequest,
};
use statespec_generated::runtime_feature_flags::RuntimeFeatureFlagStore;
use statespec_generated::runtime_leases::RuntimeLeaseStore;
use statespec_generated::runtime_logs::RuntimeLogSink;
use statespec_generated::runtime_metrics::RuntimeMetricSink;
use statespec_generated::runtime_queues::RuntimeQueueStore;
use statespec_generated::runtime_workflows::RuntimeWorkflowStore;
use statespec_generated::workflow::{
    ClaimWorkflowStepRequest, KeepAliveWorkflowStepRequest, RegisterWorkflowDefinitionRequest,
    StartWorkflowRequest, WorkflowDefinition, WorkflowStepDefinition, WorkflowStore,
};

#[test]
fn memory_backend_stores_compose_in_transaction() {
    let backend = InMemoryBackend::new();
    let flags = RuntimeFeatureFlagStore::new();
    let queues = RuntimeQueueStore::new();
    let leases = RuntimeLeaseStore::new();
    let workflows = RuntimeWorkflowStore::new();
    let logs = RuntimeLogSink::new();
    let metrics = RuntimeMetricSink::new();

    let mut tx = backend.begin().unwrap();
    <RuntimeFeatureFlagStore as FeatureFlagStore<InMemoryBackend>>::register_definition_tx(
        &flags,
        &mut tx,
        &FeatureFlagDefinition {
            name: "new_scheduler".to_string(),
            flag_type: FeatureFlagType::Bool,
            default_value: FeatureFlagValue::Bool(true),
            scope: FeatureFlagScopeKind::System,
            owner: None,
            description: None,
            expires: None,
        },
    )
    .unwrap();
    <RuntimeQueueStore as QueueStore<InMemoryBackend>>::register_definition_tx(
        &queues,
        &mut tx,
        &RegisterQueueDefinitionRequest {
            definition: QueueDefinition {
                queue: "workflow".to_string(),
                channel: "launch".to_string(),
                visibility_timeout: Duration::from_secs(60),
                max_attempts: 3,
                dead_letter_queue: None,
                metadata: Json::Object(BTreeMap::new()),
            },
        },
    )
    .unwrap();
    let lease_id = LeaseDefinitionId {
        name: "workflow_step".to_string(),
        version: 1,
    };
    <RuntimeLeaseStore as LeaseStore<InMemoryBackend>>::register_definition_tx(
        &leases,
        &mut tx,
        &LeaseDefinition {
            id: lease_id.clone(),
            resource_pattern: "workflow/*".to_string(),
            ttl: Duration::from_secs(60),
            renew_every: Duration::from_secs(30),
            max_ttl: None,
            fencing_token: true,
        },
    )
    .unwrap();
    <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::register_definition_tx(
        &workflows,
        &mut tx,
        &RegisterWorkflowDefinitionRequest {
            definition: WorkflowDefinition {
                workflow_name: "ProvisionService".to_string(),
                workflow_version: 1,
                start_step: "validate_request".to_string(),
                expected_execution_time: Duration::from_secs(60),
                singleton: false,
                steps: vec![WorkflowStepDefinition {
                    name: "validate_request".to_string(),
                    expected_execution_time: Duration::from_secs(1),
                    max_retries: 3,
                }],
                metadata: Json::Object(BTreeMap::new()),
            },
        },
    )
    .unwrap();
    <RuntimeLogSink as LogSink<InMemoryBackend>>::register_definition_tx(
        &logs,
        &mut tx,
        &LogDefinition {
            name: "launch_decision".to_string(),
            level: LogLevel::Info,
            event_name: "workflow.launch.decision".to_string(),
            fields: vec![],
        },
    )
    .unwrap();
    <RuntimeMetricSink as MetricSink<InMemoryBackend>>::register_definition_tx(
        &metrics,
        &mut tx,
        &MetricDefinition {
            name: "launch_attempts".to_string(),
            kind: MetricKind::Counter,
            backend_name: "workflow_launch_attempts_total".to_string(),
            unit: "1".to_string(),
            labels: vec![],
        },
    )
    .unwrap();
    backend.commit(tx).unwrap();

    let flag = flags
        .evaluate(
            &backend,
            &FeatureFlagEvaluationRequest {
                name: "new_scheduler".to_string(),
                context: FeatureFlagEvaluationContext::default(),
            },
        )
        .unwrap();
    assert_eq!(flag.as_bool(), Some(true));

    let now = SystemTime::UNIX_EPOCH + Duration::from_secs(100);
    let message = queues
        .enqueue(
            &backend,
            &EnqueueMessageRequest {
                message_id: "msg-1".to_string(),
                queue: "workflow".to_string(),
                channel: "launch".to_string(),
                idempotency_key: None,
                payload: Json::Object(BTreeMap::new()),
            },
        )
        .unwrap();
    assert_eq!(message.status, "Pending");
    let claimed = queues
        .claim(
            &backend,
            &ClaimMessageRequest {
                queue: "workflow".to_string(),
                channel: "launch".to_string(),
                claimant: "worker-1".to_string(),
                now,
                visibility_timeout: Duration::from_secs(60),
                max_messages: 1,
            },
        )
        .unwrap();
    assert_eq!(claimed.len(), 1);

    let lease = leases
        .acquire(
            &backend,
            &LeaseAcquireRequest {
                definition_id: lease_id,
                resource: "workflow/msg-1".to_string(),
                holder: "worker-1".to_string(),
                now,
            },
        )
        .unwrap();
    assert!(lease.acquired);

    workflows
        .start(
            &backend,
            &StartWorkflowRequest {
                workflow_execution_id: "wf-1".to_string(),
                workflow_name: "ProvisionService".to_string(),
                workflow_version: 1,
                start_step: "validate_request".to_string(),
                state: Json::Object(BTreeMap::new()),
            },
        )
        .unwrap();
    let steps = workflows
        .claim_steps(
            &backend,
            &ClaimWorkflowStepRequest {
                workflow_execution_id: "wf-1".to_string(),
                workflow_name: "ProvisionService".to_string(),
                workflow_version: 1,
                worker: "worker-1".to_string(),
                now,
                lease_duration: Duration::from_secs(60),
                max_steps: 1,
            },
        )
        .unwrap();
    assert_eq!(steps.len(), 1);
    workflows
        .keep_alive_step(
            &backend,
            &KeepAliveWorkflowStepRequest {
                workflow_execution_id: "wf-1".to_string(),
                worker: "worker-1".to_string(),
                current_step: "validate_request".to_string(),
                now,
                lease_duration: Duration::from_secs(60),
            },
        )
        .unwrap();

    logs.emit_log(
        &backend,
        &LogEvent {
            name: "launch_decision".to_string(),
            level: LogLevel::Info,
            event_name: "workflow.launch.decision".to_string(),
            fields: BTreeMap::new(),
        },
    )
    .unwrap();
    metrics
        .record_metric(
            &backend,
            &MetricSample {
                name: "launch_attempts".to_string(),
                kind: MetricKind::Counter,
                backend_name: "workflow_launch_attempts_total".to_string(),
                value: 1.0,
                unit: "1".to_string(),
                labels: BTreeMap::new(),
            },
        )
        .unwrap();
    assert_eq!(logs.inspect_events(&backend).unwrap().len(), 1);
    assert_eq!(metrics.inspect_samples(&backend).unwrap().len(), 1);

    let mut tx = backend.begin().unwrap();
    backend
        .put(
            &mut tx,
            "orders",
            "order-1",
            Json::Object(BTreeMap::from([(
                "status".to_string(),
                Json::String("new".to_string()),
            )])),
        )
        .unwrap();
    backend.commit(tx).unwrap();
    let mut tx = backend.begin().unwrap();
    assert_eq!(
        backend
            .query(
                &mut tx,
                "orders",
                &Query::KeyPrefix {
                    prefix: "order-".to_string()
                },
            )
            .unwrap()
            .len(),
        1
    );
    backend.commit(tx).unwrap();
}
